hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java

```diff
@@ -1107,20 +1108,27 @@ protected Predicate determineWhereClauseRestrictionWithJoinEmulation(AbstractUpd
 					querySpec.applyPredicate( tableReferenceJoin.getPredicate() );
 				}
 				for ( TableGroupJoin tableGroupJoin : root.getTableGroupJoins() ) {
-					assert tableGroupJoin.getJoinType() == SqlAstJoinType.INNER;
-					querySpec.getFromClause().addRoot( tableGroupJoin.getJoinedGroup() );
-					querySpec.applyPredicate( tableGroupJoin.getPredicate() );
+					if ( tableGroupJoin.getJoinType() == SqlAstJoinType.INNER ) {
```

Comment:
By avoiding the nested subquery you mean removing the left join special handling in `visitNullnessPredicate`, right?

In any case, I've tried implementing this but sadly, at least on H2, I get the error `Column "PERSON.ACCOUNT_ID" not found` for the query:
```sql
    delete 
    from
        Person 
    where
        exists(select
            1 
        from
            ACCOUNT_TABLE la1_0 
        left join
            ACCOUNT_TABLE a1_0 
                on a1_0.id=Person.account_id 
        where
            la1_0.id=Person.lazyAccount_id 
            and a1_0.id is null 
            and la1_0.code<>'aaa')
```
I guess you can't left join using a column from the target of a mutation statement.

Reply:
At least on PostgreSQL and MySQL this works and produces a better execution plan. Now, if it doesn't work on H2, then we'd indeed have to use the double nested subquery approach for H2, but I would prefer if we have a code path for "sane" databases to avoid double nesting since that comes with a bunch of other problems.

Reply:
I have added an implementation which can handle both the virtual table group + left join and the "old" roots implementation based on a new `supportsJoinInMutationStatementSubquery()` flag. H2 and Derby seem to be the only databases that complain about the subquery's join.